src/main/java/org/team1540/robot2025/RobotState.java

```diff
@@ -0,0 +1,165 @@
+                lastGyroRotation,
+                lastModulePositions,
+                Pose2d.kZero,
+                VecBuilder.fill(0.1, 0.1, 0.1),
+                VecBuilder.fill(0.5, 0.5, 5.0));
+        resetTimer.start();
+
+        SmartDashboard.putData(field);
+
+        AutoLogOutputManager.addObject(this);
+    }
+
+    public SwerveDriveKinematics getKinematics() {
+        return kinematics;
+    }
+
+    public void addOdometryObservation(SwerveModulePosition[] modulePositions, Rotation2d gyroAngle, double timestamp) {
+        lastModulePositions = modulePositions;
+        lastGyroRotation = gyroAngle;
+        poseEstimator.updateWithTime(timestamp, gyroAngle, modulePositions);
+        field.setRobotPose(getEstimatedPose());
+    }
+
+    public void addVisionMeasurement(PoseObservation visionPose) {
+        if (shouldAcceptVision(visionPose)) {
+            poseEstimator.addVisionMeasurement(
+                    visionPose.estimatedPoseMeters().toPose2d(),
+                    visionPose.lastMeasurementTimestampSecs(),
+                    getStdDevs(visionPose));
+        }
+    }
+
+    private Matrix<N3, N1> getStdDevs(PoseObservation poseObservation) {
+        double xyStdDev =
+                XY_STD_DEV_COEFF * Math.pow(poseObservation.avgTagDistance(), 2.0) / poseObservation.numTagsSeen();
+        double rotStdDev =
+                ROT_STD_DEV_COEFF * Math.pow(poseObservation.avgTagDistance(), 2.0) / poseObservation.numTagsSeen();
+        boolean acceptYaw =
+                poseObservation.numTagsSeen() > 1 || (poseObservation.numTagsSeen() > 0 && DriverStation.isDisabled());
+        return VecBuilder.fill(xyStdDev, xyStdDev, acceptYaw ? rotStdDev : Double.POSITIVE_INFINITY);
+    }
+
+    private boolean shouldAcceptVision(PoseObservation poseObservation) {
+        Pose3d estimatedPose = poseObservation.estimatedPoseMeters();
+        return poseObservation.numTagsSeen() >= MIN_ACCEPTED_NUM_TAGS // Must see sufficient tags
+                && poseObservation.avgTagDistance() <= MAX_ACCEPTED_AVG_TAG_DIST_METERS // Must be close enough
+                // Must be within field roughly
+                && estimatedPose.getX() >= -MAX_OUTSIDE_OF_FIELD_TOLERANCE
+                && estimatedPose.getX() <= APRIL_TAG_FIELD_LAYOUT.getFieldLength() + MAX_OUTSIDE_OF_FIELD_TOLERANCE
+                && estimatedPose.getY() >= -MAX_OUTSIDE_OF_FIELD_TOLERANCE
+                && estimatedPose.getY() <= APRIL_TAG_FIELD_LAYOUT.getFieldWidth() + MAX_OUTSIDE_OF_FIELD_TOLERANCE
+                // Must not be actively flying
+                && estimatedPose.getZ() <= MAX_ROBOT_Z_TOLERANCE
+                // Must not be translating or rotating too fast
+                && Math.abs(robotVelocity.omegaRadiansPerSecond) <= MAX_ACCEPTED_ROT_SPEED_RAD_PER_SEC
+                && Math.hypot(robotVelocity.vxMetersPerSecond, robotVelocity.vyMetersPerSecond)
+                <= MAX_ACCEPTED_LINEAR_SPEED_MPS;
+    }
```

Comment:
I'm not sure why, but in my simulation testing this method is always returning false, so vision poses are never being accepted.

Reply:
OK I have identified your issue (kind of). If you remove the max tag distance requirement here then it works. Also, it's probably unnecessary since you already account for tag distance with the stddevs, so you're just throwing away otherwise useful information here. Also, since our cameras this year are global shutter cameras the limit based on velocities is probably also unneeded.